### PR TITLE
Add support for OTP 27

### PR DESCRIPTION
**Pull request by kivra-pauoli**
👋 

I'm not sure what the support policy/maintenance range for `brod` is, but one could already start moving toward OTP 27 for systems that are adopting it.

I tested these changes in my fork and CI was Ok using Erlang/OTP 27 and `rebar3` 3.23: https://github.com/kivra-pauoli/brod/actions/runs/9564503494/job/26365813507

I'll add comments on top of the changes to ease review. I tried to do commit-by-commit so it's easier to do so too, in any case.

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {deps, [ {kafka_protocol, "4.1.5"}
        , {snappyer, "1.2.9"}
        ]}.
-{project_plugins, [{coveralls, "~> 2.2.0"}, {rebar3_lint, "~> 1.0.2"}]}.
+{project_plugins, [{coveralls, "~> 2.2.0"}, {rebar3_lint, "~> 3.2.5"}]}.
```
Comment (kivra-pauoli) [inline rebar.config:4]: `rebar3_lint` depends on `katana_code` that would fail when imported in `1.0.2`.

File: rebar.config [hunk 2]
```diff
@@ -28,7 +28,7 @@
            , {jsone, "1.7.0"}
            , {meck, "0.9.2"}
            , {proper, "1.4.0"}
-           , {snabbkaffe, {git, "https://github.com/kafka4beam/snabbkaffe.git", {branch, "1.0.8"}}}
+           , {snabbkaffe, {git, "https://github.com/kafka4beam/snabbkaffe.git", {branch, "1.0.10"}}}
```
Comment (kivra-pauoli) [inline rebar.config:31]: This update was required because the dep. was failing in OTP 27.

File: rebar.lock
Comment (kivra-pauoli) [inline rebar.lock:1]: This is `rebar3 do unlock --all, upgrade --all`, as-is. At the moment `rebar.config` (in `brod`) states one thing, `rebar.lock` states another thing.

File: src/brod_producer.erl
Comment (kivra-pauoli) [inline src/brod_producer.erl:1]: Motivation for this change (apart from the fact we're fixing a compilation bug) is found in the commit message: https://github.com/kafka4beam/brod/pull/585/commits/cf3bd74a9166c3504ff20c332f5f0ba14a3a2903

File: src/brod_supervisor3.erl
Comment (kivra-pauoli) [inline src/brod_supervisor3.erl:1]: Motivation for this change (apart from the fact we're fixing a compilation bug) is found in the commit message: https://github.com/kafka4beam/brod/pull/585/commits/cf3bd74a9166c3504ff20c332f5f0ba14a3a2903

File: src/brod_producer.erl [hunk 3]
```diff
@@ -36,7 +36,7 @@
         , handle_info/2
         , init/1
         , terminate/2
-        , format_status/2
```
Comment (zmstone) [inline src/brod_producer.erl:39]: please keep `format_status/2` for older version OTP

**Comment by zmstone** [comment]
Please add OTP 27 to the ci matrix.

File: src/brod_producer.erl (hunk 3 above)
Comment (kivra-pauoli) [reply src/brod_producer.erl:39]: To do this we need to remove `warnings_as_errors` from the compilation options, as Erlang/OTP complains then the process stops. I'll push it, reversing the change, so you can check what happens.

Comment (kivra-pauoli) [reply src/brod_producer.erl:39]: I can add an OTP_RELEASE exception, if that's what you mean.

**Comment by kivra-pauoli** [comment]
> Please add OTP 27 to the ci matrix.

CI presents two distinct scenarios:

1. env variables determine a specific OTP/Rebar3 combo
2. CI matrix determines a combo of stuff to test on

I'm gonna update the matrix, but I don't know how you prefer to deal with the env variables being fixed

File: src/brod_producer.erl (hunk 3 above)
Comment (kivra-pauoli) [reply src/brod_producer.erl:39]: b38712d

**Comment by zmstone** [comment]
sorry, it seems the erlef/setup-beam@v1 has rebar3 built on otp 26.
not sure how to fix, you can remove otp 27 from the ci yaml for now if you wish.

**Comment by kivra-pauoli** [comment]
> sorry, it seems the erlef/setup-beam@v1 has rebar3 built on otp 26.

It's typically build on top of `current`-2, so that'd be 25. If I didn't pin the rebar3 executable correctly I will (**edit**: 3.23.0 should do it - check 42a954b).

**Review by zmstone — APPROVED** [review]

**Comment by kivra-pauoli** [comment]
Thanks much. Do you have an ETA on the next release that includes this change?

**Comment by zmstone** [comment]
included in 3.19.1

**Pull request merged**